linode/instanceconfig/resource.go

```diff
@@ -271,22 +272,52 @@ func updateResource(ctx context.Context, d *schema.ResourceData, meta any) diag.
 		shouldUpdate = true
 	}
 
+	powerOffRequired := false
 	if d.HasChange("interface") {
-		putRequest.Interfaces = helper.ExpandInterfaces(ctx, d.Get("interface").([]any))
+		putRequest.Interfaces = helper.ExpandConfigInterfaces(ctx, d.Get("interface").([]any))
+		config, err := client.GetInstanceConfig(ctx, linodeID, id)
+		if err != nil {
+			return diag.Errorf("failed to get config %d: %s", id, err)
+		}
+
+		bootedConfigID, err := helper.GetCurrentBootedConfig(ctx, &client, linodeID)
+		if err != nil {
+			tflog.Warn(
+				ctx, fmt.Sprintf("failed to get current booted config of Linode %d", linodeID),
+			)
+		}
+
+		powerOffRequired = instancehelpers.VPCInterfaceIncluded(config.Interfaces, putRequest.Interfaces) && id == bootedConfigID
 		shouldUpdate = true
 	}
 
+	// We should not use `HasChange(...)` here because of possible mid-apply changes
+	managedBoot := !d.GetRawConfig().GetAttr("booted").IsNull()
+
+	shouldPowerBackOn := !managedBoot && powerOffRequired && inst.Status == linodego.InstanceRunning
+
 	if shouldUpdate {
+		if powerOffRequired {
+			instancehelpers.ShutdownInstanceForVPCInterfaceUpdate(
+				ctx, meta.(*helper.ProviderMeta), linodeID, helper.GetDeadlineSeconds(ctx, d),
+			)
+		}
+
 		tflog.Debug(ctx, "Update detected, sending config PUT request to API", map[string]any{
 			"body": putRequest,
 		})
-		if _, err := client.UpdateInstanceConfig(ctx, linodeID, int(id), putRequest); err != nil {
+		if _, err := client.UpdateInstanceConfig(ctx, linodeID, id, putRequest); err != nil {
 			return diag.Errorf("failed to update instance config: %s", err)
 		}
+
+		if shouldPowerBackOn {
+			instancehelpers.BootInstanceAfterVPCInterfaceUpdate(
+				ctx, meta.(*helper.ProviderMeta), linodeID, id, helper.GetDeadlineSeconds(ctx, d),
+			)
+		}
 	}
 
-	// We should not use `HasChange(...)` here because of possible mid-apply changes
-	if !d.GetRawConfig().GetAttr("booted").IsNull() {
+	if managedBoot {
 		if err := applyBootStatus(ctx, &client, inst, id, helper.GetDeadlineSeconds(ctx, d),
 			d.Get("booted").(bool)); err != nil {
```

Comment:
This seems to be hanging when `booted = true` and the interfaces are updated to add or drop a VPC. I _think_ this is because the instance is shutdown [here](https://github.com/linode/terraform-provider-linode/pull/1265/files#diff-d0c91e4235901788aabc4f154c3636efb3ad330d1091f3616df04d47b9b0c565R301) but the instance stored in `inst` isn't updated, so the `applyBootStatus` function assumes the instance is already booted and hangs [here](https://github.com/linode/terraform-provider-linode/blob/ca53a045daa502e389a990ee4f65f7627a279812/linode/instanceconfig/helper.go#L196-L201).

As a hacky fix, you could add `inst.Status = linodego.InstanceOffline` and `inst.Status = linodego.InstanceRunning` on lines 304 and 317 respectively. Alternatively you could just refresh the instance right before this call 👍

Reply:
I think it might be better to move instance get call to `applyBootStatus` function so it can guarantee itself always work as expected. Do you think that's a good idea?